domain/secretbackend/state/state.go

```diff
@@ -46,66 +47,253 @@ func NewState(factory coredatabase.TxnRunnerFactory, logger Logger) *State {
 // model is found the given uuid then an error of
 // [github.com/juju/juju/domain/model/errors.NotFound] is returned.
-func (s *State) GetModel(ctx context.Context, uuid coremodel.UUID) (coremodel.Model, error) {
+func (s *State) GetModel(ctx context.Context, uuid coremodel.UUID) (secretbackend.Model, error) {
 	db, err := s.DB()
 	if err != nil {
-		return coremodel.Model{}, errors.Trace(err)
+		return secretbackend.Model{}, errors.Trace(err)
 	}
 
-	var m coremodel.Model
-	return m, db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		var err error
-		m, err = modelstate.Get(ctx, tx, uuid)
-		return err
+	stmt, err := s.Prepare(`
+SELECT &Model.*
+FROM v_model_secret_backend
+WHERE uuid = $M.uuid`, sqlair.M{}, Model{})
+	if err != nil {
+		return secretbackend.Model{}, errors.Trace(err)
+	}
+	var m Model
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, sqlair.M{"uuid": uuid}).Get(&m)
+		if errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("%w: %q", modelerrors.NotFound, uuid)
+		}
+		return errors.Trace(err)
 	})
+	if err != nil {
+		return secretbackend.Model{}, errors.Trace(domain.CoerceError(err))
+	}
+	if !m.Type.IsValid() {
+		// This should never happen.
+		return secretbackend.Model{}, fmt.Errorf("invalid model type for model %q", m.Name)
+	}
+	// If the backend ID is NULL, it means that the model does not have a secret backend configured.
+	// We return an empty string in this case.
+	// TODO: we should return the `default` backend once we start to include the
+	// `internal` and `k8s` backends in the database. And obviously, this field will become non-nullable.
+	return secretbackend.Model{
+		ID:              m.ID,
+		Name:            m.Name,
+		Type:            m.Type,
+		SecretBackendID: m.SecretBackendID.String,
+	}, nil
 }
 
 // UpsertSecretBackend persists the input secret backend entity.
 func (s *State) UpsertSecretBackend(ctx context.Context, params secretbackend.UpsertSecretBackendParams) (string, error) {
+	if err := params.Validate(); err != nil {
+		return "", errors.Trace(err)
+	}
+
 	db, err := s.DB()
 	if err != nil {
 		return "", errors.Trace(err)
 	}
-	operation := upsertOperation{UpsertSecretBackendParams: params}
-	if err := operation.build(); err != nil {
+
+	getBackendStmt, err := s.Prepare(`
+SELECT &SecretBackend.*
+FROM secret_backend
+WHERE uuid = $M.uuid`, SecretBackend{}, sqlair.M{})
+	if err != nil {
 		return "", errors.Trace(err)
 	}
+
+	upsertBackendStmt, err := s.Prepare(`
+INSERT INTO secret_backend
+    (uuid, name, backend_type, token_rotate_interval)
+VALUES (
+    $SecretBackend.uuid,
+    $SecretBackend.name,
+    $SecretBackend.backend_type,
+    $SecretBackend.token_rotate_interval
+)
+ON CONFLICT (uuid) DO UPDATE SET
+    name=EXCLUDED.name,
+    token_rotate_interval=EXCLUDED.token_rotate_interval;`, SecretBackend{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	upsertRotationStmt, err := s.Prepare(`
```

Comment:
We've been putting the code to write the related tables into separate helper methods, see cloud config or credential config storage pool config etc. It keeps the top level method much tidier.